lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,119 @@
+    exports.SpellChecker = class SpellChecker
         # By now it's just a stub, but I guess this function will take data from server somehow in future.
         # TODO
         # @return {Object} JSON-list of incorrect words.
         getJson: =>
             return TEST_JSON_TYPOS
 
-        # Function checks whether token is in JSON incorrect words list.
-        # @param {String} token Token to check.
+        # Check whether token is in JSON incorrect words list.
+        # @param {String} token: Token to check.
         # @return {Boolean} False if token is in list, true otherwise.
         check: (token) =>
-            wordsList = @getJson()
-            for item in wordsList
-                if token == item.word
-                    console.log(token, item.corrections)    # Let's just put this out to console for now
-                    return false
-            return true
+            correctionsList = @getJson()
+            # New json structure makes it possible to search for token just like this
+            return !correctionsList[token]
+
+    # Implements some routines to show popup for spellchecker (to choose a proper substitution for a typo).
+    # Also binds popup to an editor's shortcut.
+    exports.setupSpellCheckerPopup = (editor) ->
+        # Take corrections for current word and shows them in a popup.
+        # @param {Array} corrections: popup has a setData function that takes some sort of
+        # structure, it needs to be replicated so popup could show options for substitution.
+        #
+        # Some of this code is duplicated from autocomplete.js so it's just left as-is (by now?)
+        showPopup = (options) =>
+            AcePopup = require("ace/autocomplete/popup").AcePopup
+            util = require("ace/autocomplete/util")
+            HashHandler = require("ace/keyboard/hash_handler").HashHandler
+            event = require("ace/lib/event")
+
+            @popup = new AcePopup(document.body ? document.documentElement)
+            @popup.setData(options)
+            @popup.setTheme(editor.getTheme())
+            @popup.setFontSize(editor.getFontSize())
+
+            @commands = {
+                # TODO
+                # Doesn't work yet for some reason
+                "Esc": (editor) ->
+                    @popup.hide()
+                    editor.keyBinding.removeKeyboardHandler(@keyboard);
+                    console.log("esc")
+
+            }
+            @keyboard = new HashHandler()
+            @keyboard.bindKeys(@commands)
+            editor.keyBinding.addKeyboardHandler(@keyboard)
+
+            # This part is like "magic, do not touch" thing. Was copied from autocomplete.js,
+            # seems to set things up in a certain way so popup window opens under the cursor.
+            renderer = editor.renderer
+            lineHeight = renderer.layerConfig.lineHeight
+            pos = renderer.$cursorLayer.getPixelPosition(this.base, true)
```

Comment:
I guess you can find the answer in debugger, but the real issue is that there might be no `base` in whatever `this` is now. So you'd better check.